src/command.ml

```diff
@@ -148,15 +148,42 @@ let string_of_command_spec_with_calls call_with_tags call_with_target resolve_vi
       | P p -> put_space (); put_filename p
       | Px u -> put_space (); put_filename u; call_with_target u
       | V v -> if resolve_virtuals then do_spec (virtual_solver v)
-               else (put_space (); Printf.bprintf b "<virtual %s>" (Shell.quote_filename_if_needed v))
-      | S l -> List.iter do_spec l
+              else (put_space (); Printf.bprintf b "<virtual %s>" (encode v))
+      | S l -> do_specs l
       | T tags -> call_with_tags tags; do_spec (!tag_handler tags)
-      | Quote s -> put_space (); put_filename (aux s)
+      | Quote s -> put_space (); put_filename (aux encode s)
+    and do_specs = function
+      | [] -> ()
+      | (A "-pp" as pp) :: Quote q :: rest ->
+        do_spec pp;
+        let c =
+          if Sys.win32 then
+            (* external commands are unfortunately called called with cmd.exe
+               (via Sys.command).
+               Cmd.exe has strange quoting rules. The most notorious quirk is, that
+               you can't use forward slashes as path separators at the first position,
+               unless you quote the expression explicitly.
+               cmd.exe will interpret the slash and everything thereafter as first
+               parameter. Eg. 'bin/foo -x' is treated like 'bin /foo -x'. *)
+            let first = ref true in
+            aux (fun x ->
+                if ((!first && (first := false; String.contains x '/')) || not (Shell.is_simple_filename x))
+                then (My_std.quote_cmd ("\"" ^ x ^ "\""))
+                else x) q
```

Comment:
The logic here is fairly complex and also not clearly correct: if `q` is to be interpreted as a command, then we need different quoting rules for the first argument of `q`, which is *not* necessarily an argument on which the `encode` function will be called. So `encode` will run on a follow-up argument and the logic does not match the intent.

An alternative proposal would be as follows: on Windows, ensure that `is_simple_filename` fails on filenames that contain a `/`. Does that suffice to prevent the issue that you are guarding against here? (I would be happy to keep your nice, descriptive comment that explains why we need to protect certain occurrences of `/`, which could move to Shell if we do it this way.)

Reply:
I don't like moving this to Shell because it should only apply inside cmd.
I've change the logic to apply no only on the first encoded element.

Reply:
Remember that windows has more constraint on command line size.